zict/cache.py

```diff
@@ -44,44 +42,75 @@ class Cache(ZictBase[KT, VT]):
     data: MutableMapping[KT, VT]
     cache: MutableMapping[KT, VT]
     update_on_set: bool
+    _gen: int
+    _last_updated: dict[KT, int]
 
     def __init__(
         self,
         data: MutableMapping[KT, VT],
         cache: MutableMapping[KT, VT],
         update_on_set: bool = True,
     ):
+        super().__init__()
         self.data = data
         self.cache = cache
         self.update_on_set = update_on_set
+        self._gen = 0
+        self._last_updated = {}
 
+    @locked
     def __getitem__(self, key: KT) -> VT:
         try:
             return self.cache[key]
         except KeyError:
             pass
-        value = self.data[key]
-        self.cache[key] = value
-        return value
+        gen = self._last_updated[key]
 
-    def __setitem__(self, key: KT, value: VT) -> None:
-        # If the item was already in cache and data.__setitem__ fails, e.g. because it's
-        # a File and the disk is full, make sure that the cache is invalidated.
-        try:
-            del self.cache[key]
-        except KeyError:
-            pass
+        with self.unlock():
+            value = self.data[key]
 
-        self.data[key] = value
-        if self.update_on_set:
+        # Could another thread have called __setitem__ or __delitem__ on the
+        # same key in the meantime? If not, update the cache
+        if gen == self._last_updated.get(key):
             self.cache[key] = value
+            self._last_updated[key] += 1
+        return value
 
+    @locked
+    def __setitem__(self, key: KT, value: VT) -> None:
+        # If the item was already in cache and data.__setitem__ fails, e.g. because
+        # it's a File and the disk is full, make sure that the cache is invalidated.
+        discard(self.cache, key)
+        gen = self._gen
+        gen += 1
+        self._last_updated[key] = self._gen = gen
+
+        with self.unlock():
```

Comment:
Can we comment here why this benefits from being unlocked? It seems like some of the second condition below can be removed if this were left locked. Assume there is a reason for the trade-off though; to allow another thread to make progress? If so, I'm curious if it's worth the tradeoff for added complexity of checking / correcting state (second conditional call to `discard(self.cache, key)`) in that same second conditional.

Reply:
It's explained in the notes at the top of every class.
These methods can take arbitrarily long to run, so two threads should be able to run in unison on the same mapping while they're running:

- `Buffer.slow.__setitem__`
- `Buffer.slow.__getitem__`
- `LRU.on_evict`
- `Cache.data.__setitem__`
- `Cache.data.__getitem__`
- `File.__setitem__` (specifically fh.write / fh.writelines; not the other syscalls)
- `File.__getitem__` (specifically fh.readinto; not the other syscalls)

All other methods are expected to be _fast_, which should be intended as _it's OK for a thread that must not be busy_ (read: the Worker's event loop) _to lock waiting for them to finish_.